provider/adapters/src/fork/by_error.rs

```diff
@@ -0,0 +1,263 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::marker::PhantomData;
+
+use super::ForkByErrorPredicate;
+use alloc::vec::Vec;
+#[cfg(feature = "datagen")]
+use icu_provider::datagen;
+use icu_provider::prelude::*;
+
+/// A provider that returns data from one of two child providers based on a predicate function.
+///
+/// This is an abstract forking provider that must be provided with a type implementing the
+/// [`ForkByErrorPredicate`] trait.
+#[derive(Debug, PartialEq, Eq)]
+pub struct ForkByErrorProvider<P0, P1, F>(pub P0, pub P1, PhantomData<F>);
```

Comment:
Nit: put a bound on `F`

Reply:
It's standard rust style to avoid bounds on structs unless absolutely necessary (due to associated types), instead using impls, and, more in general, only using the bounds you need.

In the case of structs it's in part because otherwise things get super unnecessarily infectious, it's quite annoying.